Performance enhancements for JSON printing

SeanTAllen opened the pull request:
Pretty printing JSON creates a slightly larger data structure than
a non-pretty printed version. When dealing with large aggregates,
this can result in large differences that can have a noticeable
impact on performance when sending over the network.

This change makes pretty printing optional in a way that maintains
compatibility with the existing API.

jemc commented on packages/json/tests.pony:
```diff
@@ -532,6 +572,32 @@ class iso _TestPrintObject is UnitTest
+    obj.data("b") = I64(3)
+    let s = doc.string(false)
+    h.assert_true((s == "{\"a\": true, \"b\": 3}") or
+      (s == "{\"b\": false, \"a\": true}"))
```
This clause wouldn't ever be true, right?  It should be `"b": 3`, not `"b": false`.

jemc commented:
I would expect a non-pretty JSON string to not include spaces after commas and colons.

jemc commented:
Also, I wonder if it makes sense to make not-pretty the default?

As far as I can recall, making the most compact string possible is usually the default on all JSON libraries I've dealt with, with a pretty-printed string being the exception, not the rule.

I don't think it's really a compatibility issue, since either way we're generating valid JSON that could be consumed later.  I think it's more important to be performant-by-default, so that users who don't think about the pretty-printing aspect (and may never visually inspect their strings) are getting the best performance by default.  If the user is visually inspecting their strings, and wants a pretty version to inspect, I think that's the case where they should be specifying a preference for prettiness.

SeanTAllen commented:
All excellent points @jemc. I'll make changes and push sometime this weekend or early next week.

SeanTAllen commented:
ugh. the tests were testing the exact formatting. ugh.

SeanTAllen commented:
fixed tests and pushed

jemc commented on CHANGELOG.md:
```diff
@@ -82,6 +82,7 @@ All notable changes to the Pony compiler and standard library will be documented
 - pony_alloc_msg_size runtime function
 - `net/WriteBuffer`
 - `serialise` package.
+- optional parameter in json objects, arrays, and doc to turn off pretty printing
```
This should now read "optional parameter [...] to turn *on* pretty printing".

jemc commented on packages/json/json.pony:
```diff
@@ -75,7 +85,17 @@ class JsonObject
+          _JsonPrint._string(i._2, elem_indent, true)
+      else
+        text = text + if text.size() > 0 then " " else "" end
+          + "\"" + i._1 + "\": " +
```
The space-after-colon should be removed in not-pretty printing to save the byte.

jemc commented on packages/json/json.pony:
```diff
@@ -75,7 +85,17 @@ class JsonObject
+        text = text + "\n" + elem_indent + "\"" + i._1 + "\": " +
+          _JsonPrint._string(i._2, elem_indent, true)
+      else
+        text = text + if text.size() > 0 then " " else "" end
```
This entire conditional expression can be removed to remove the space-after-comma to save extra bytes when not-pretty printing.

jemc commented:
While we're on the subject of JSON printing and performance, it seems like some of these string-joining expressions using `+` could be cleaned up to use `push` and `append` instead, as per [the pony pattern on limiting string allocations](http://patterns.ponylang.org/performance/limiting-string-allocations.html).

This could be in another future PR if you prefer to leave it out of this one, though.

SeanTAllen commented:
I've pushed everything except for addressing the string concatenation issue.

I'm pausing for a moment to consider...

Even with appends, because of the recursive nature of how printing works, there's going to be several allocations. I feel like there has to be a better way to structure code like this in order to minimize allocations. 

Ideally, allocate a buffer, have an object that takes the json (or a combination of objects) all of which work on the Json representation and write directly to the buffer rather than allocating a series of Strings that need to be joined together.

Anyway, back to addressing issues with + in the current design. .

If anyone has thoughts.. @ponylang/core @ponylang/committer, I'm all ears.

SeanTAllen commented:
And the travis failure is scary:

```
{"a":true,"b":3}
1 test started, 0 complete: encode/Base64.encode started
1 test started, 1 complete: encode/Base64.encode complete
2 tests started, 1 complete: encode/Base64.decode started
2 tests started, 2 complete: encode/Base64.decode complete
3 tests started, 2 complete: encode/Base64.encodedecode started
3 tests started, 3 complete: encode/Base64.encodedecode complete
4 tests started, 3 complete: encode/Base64.quote started
4 tests started, 4 complete: encode/Base64.quote complete
make: *** [test] Segmentation fault: 11
```

SeanTAllen commented:
@jemc thoughts on taking:

    `var text = recover String end`

and picking some value and starting from that as the initial allocation size like...

   ` var text = recover String(128) end`

With how this code is structured, we can't know the size ahead of time but we could pick a value as handwave idea. I'm thinking 128 for now because ¯\_(ツ)_/¯.

SeanTAllen commented:
@sylvanc picked 256 as the number. 

i'm going to push that for now.

jemc commented:
:+1: for this PR.

Regarding restructuring this to use `append` instead of `+` in a way that works recursively, it will take some refactoring.

For a first step, I would say ignore the extra allocations made for the number and string elements - let those extra allocations happen for now, and use `append` to add them to your larger running buffer.  It's easier to deal with the objects and arrays' printing first, since we control the definition here in the `json` package.

How I would approach refactoring them:
1. Move the current contents of the `string` method into a `_show` method.
2. The `_show` method should take all the current arguments that the `string` method takes, plus an existing `String iso` buffer to write into.
3. The `string` method should allocate a new `String` of size `256` and pass it to the `_show` method in a tail call.
4. When a `_show` method gets to an inner JsonObject or JsonArray element, it calls the `_show` method and passes the current running buffer.  The `_show` method would also need to return the consumed `String iso` buffer to pass it back to the caller's control.

Once this is dealt with, we can figure out how to also print the inner elements without extra allocations, but I would treat this as a separate, more invasive step.

Perelandric commented:
@jemc @SeanTAllen If no one is working on it, I can implement writing to the shared buffer if you folks would like.

SeanTAllen commented:
@Perelandric have at it!
